compress.cc

```diff
@@ -102,7 +113,34 @@ compressor::ptr_type compressor::create(const sstring& name, const opt_getter& o
         }
     }
 
-    return compressor_registry::create(qn, opts);
+    static thread_local std::unordered_map<sstring, ptr_type> named_compressors;
+
+    ptr_type res = nullptr;
+
+    auto i = named_compressors.find(name);
+    if (i != named_compressors.end()) {
+        res = i->second;
+        /**
+         * If we have a cached compressor, we need to potentially
+         * (shallow) replace it to ensure we respect the compressor
+         * parameters set by "opts" (or lack thereof).
+         * Implementations can handle this in "replace", and either
+         * make an expensive replacement, or, more likely, a very cheap one...
+         * (zstd).
+        */
+        auto rep = res->replace(opts);
+        if (rep) {
+            res = std::move(rep);
```

Comment:
Doesn't this mean that we create a new compressor every time we need to use zstd with a different compression level than last time?

Imagine there are two actively-used tables (each with 50% of the traffic), which are compressed with different zstd levels. Won't every sstable read have a ~50% chance to create a new compressor?

Reply:
Yes. Which is 50% less than the chance is atm, since zstd is never cached in existing code. However, in this case the "replace" part gets into action, and in the case of zstd, we still share the relevant zctxt objects, only the front-end object is unique. Which is neglible, and relatively shortlived.

Reply:
Right, nevermind.